sql/src/main/java/io/crate/execution/engine/indexing/UpsertResultCollectors.java

```diff
@@ -0,0 +1,153 @@
+import java.util.List;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.BinaryOperator;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collector;
+
+public final class UpsertResultCollectors {
+
+    public static Collector<ShardUpsertRequestAndResponse, UpsertResults, Iterable<Row>> newRowCountCollector() {
+        return new RowCountCollector();
+    }
+
+    public static Collector<ShardUpsertRequestAndResponse, UpsertResults, Iterable<Row>> newSummaryCollector(DiscoveryNode localNode) {
+        return new SummaryCollector(ImmutableMap.of(
+            "id", new BytesRef(localNode.getId()),
+            "name", new BytesRef(localNode.getName())
+        ));
+    }
+
+    private static class RowCountCollector implements Collector<ShardUpsertRequestAndResponse, UpsertResults, Iterable<Row>> {
+
+        private final Object lock = new Object();
+
+        @Override
+        public Supplier<UpsertResults> supplier() {
+            return UpsertResults::new;
+        }
+
+        @Override
+        public BiConsumer<UpsertResults, ShardUpsertRequestAndResponse> accumulator() {
+            return this::processShardResponse;
+        }
+
+        @Override
+        public BinaryOperator<UpsertResults> combiner() {
+            return (i, o) -> {
+                synchronized (lock) {
+                    o.addResult(i.getSuccessRowCountForNoUri());
+                }
+                return o;
+            };
+        }
+
+        @Override
+        public Function<UpsertResults, Iterable<Row>> finisher() {
+            return r -> Collections.singletonList(new Row1(r.getSuccessRowCountForNoUri()));
+        }
+
+        @Override
+        public Set<Characteristics> characteristics() {
+            return Sets.newHashSet(Characteristics.CONCURRENT);
+        }
+
+        void processShardResponse(UpsertResults upsertResults, ShardUpsertRequestAndResponse requestAndResponse) {
+            synchronized (lock) {
+                upsertResults.addResult(requestAndResponse.shardResponse.successRowCount());
+            }
+        }
+    }
+
+    private static class SummaryCollector implements Collector<ShardUpsertRequestAndResponse, UpsertResults, Iterable<Row>> {
+
+        private final ImmutableMap<String, BytesRef> nodeInfo;
+
+        private final Object lock = new Object();
+
+        SummaryCollector(ImmutableMap<String, BytesRef> nodeInfo) {
+            this.nodeInfo = nodeInfo;
+        }
+
+        @Override
+        public Supplier<UpsertResults> supplier() {
+            return () -> new UpsertResults(nodeInfo);
+        }
+
+        @Override
+        public BiConsumer<UpsertResults, ShardUpsertRequestAndResponse> accumulator() {
+            return this::processShardResponse;
+        }
+
+        @Override
+        public BinaryOperator<UpsertResults> combiner() {
+            return (i, o) -> {
+                synchronized (lock) {
```

Comment:
This is using the same lock for merging and receiving the shard responses. Perhaps use `synchronized(o)` here?

Reply:
afaik both must lock on the same object, but yes could lock on the upserResult in both cases..

Reply:
no this won't work afaik as it would synchronize on a local var which makes no sense

Reply:
Why does it have to use the same lock? There is not interleaving of the two operations. Locking on `o` should work since this is always the same object where results are combined.

Reply:
> no this won't work afaik as it would synchronize on a local var which makes no sense

I think the problem would be the current `combiner` implementation. If it `synchronized` only on `o` then the `i.getSuccessRowCountForNoUri` could return a stale read. I'm not sure if synchronizing on both could lead to a deadlock.
So a `lock` is probably easier and safe